### PR TITLE
Change default program type to EBPF_PROGRAM_TYPE_UNSPECIFIED

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

This PR is part1 of the changes for Issue #223.

This PR contains the following changes:

1. Changed `get_program_type_windows()` API behavior to not set the default program type to XDP.
2. `ebpf_api_elf_verify_section_from_file` now takes an optional program type as input.
3. Few bug fixes.
4. Other changes needed due to the above API changes.

## Testing

Added new tests. Existing tests and new added tests cover the changes done.

## Documentation

NA

### CODE REVIEW

File: include/ebpf_api.h [hunk 1]
```diff
@@ -184,9 +189,10 @@ extern "C"
      */
     uint32_t
     ebpf_api_elf_verify_section_from_memory(
-        const char* data,
+        _In_z_ const char* data,
```
Comment (dthaler) [inline include/ebpf_api.h:192]:
```suggestion
        _In_reads_(data_length) const char* data,
```

File: libs/api_common/windows_platform_common.cpp [hunk 2]
```diff
@@ -56,7 +58,28 @@ get_program_type_windows(const std::string& section, const std::string&)
         }
     }
 
-    return windows_xdp_program_type;
+    // Note: Ideally this function should throw exception whenever a matching ProgramType is not found,
```
Comment (dthaler) [inline libs/api_common/windows_platform_common.cpp:61]:
```suggestion
    // Note: Ideally this function should throw an exception whenever a matching ProgramType is not found,
```

File: libs/service/verifier_service.cpp [hunk 3]
```diff
@@ -49,12 +49,19 @@ verify_byte_code(
...
 {
+    std::ostringstream error;
     const ebpf_platform_t* platform = &g_ebpf_platform_windows_service;
     std::vector<ebpf_inst> instructions{instruction_array, instruction_array + instruction_count};
     program_info info{platform};
     std::string section;
     std::string file;
-    info.type = get_program_type_windows(*program_type);
+    try {
+        info.type = get_program_type_windows(*program_type);
```
Comment (dthaler) [inline libs/service/verifier_service.cpp:59]: question (not for this PR): should we change the prototype so this needn't throw an exception?

Comment (saxena-anurag) [reply libs/service/verifier_service.cpp:59]: Agree. `get_program_type_windows` today returns EbpfProgramType. If we change the prototype to return EbpfProgramType *, then this function can simply return null.

We can today also not throw an exception and instead return the EbpfProgramType for EBPF_PROGRAM_TYPE_UNSPECIFIED which will have a context descriptor with all 0s, but I don't want verifier to proceed and try to verify the program using that all zeroed out context descriptor.

**Review by dthaler — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Review by dthaler — DISMISSED** [review]

File: include/ebpf_structs.h [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include <stdbool.h>
 #include <stdint.h>
 #endif
+#include "bpf_structs.h"
```
Comment (dthaler) [inline include/ebpf_structs.h:13]: what is the difference between bpf_structs.h vs ebpf_structs.h?  I.e., when should something be added to one vs the other?  Can you add a // comment explaining?  E.g., why is XDP_FLAGS_REPLACE defined in this file and not the other one?  What's the rule?

File: libs/api/ebpf_api.cpp [hunk 5]
```diff
@@ -3197,27 +3186,42 @@ ebpf_result_t
...
+    ebpf_program_type_t* program_type_uuid;
     EBPF_LOG_ENTRY();
     ebpf_assert(name);
     ebpf_assert(program_type);
     ebpf_assert(expected_attach_type);
 
-    EbpfProgramType type = get_program_type_windows(name, name);
-    ebpf_program_type_t* program_type_uuid = (ebpf_program_type_t*)type.platform_specific_data;
+    try {
+        EbpfProgramType type = get_program_type_windows(name, name);
```
Comment (dthaler) [inline libs/api/ebpf_api.cpp:3197]: fir consistency with line 3220 can this be
```suggestion
        const EbpfProgramType& type = get_program_type_windows(name, name);
```
?

File: include/ebpf_structs.h (hunk 4 above)
Comment (saxena-anurag) [reply include/ebpf_structs.h]: Actually the new file is not needed. Removed it now.

File: libs/api/ebpf_api.cpp (hunk 5 above)
Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp]: fixed.

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**